### PR TITLE
[CPDLP-1534] Providers can trigger a clawback via the API

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

- Ticket: https://dfedigital.atlassian.net/browse/CPDLP-1534

### Changes proposed in this pull request

* Added extension to `Finance::ClawbackDeclaration` to make clawback when the declaration is paid.

### CODE REVIEW

File: app/services/void_participant_declaration.rb [hunk 1]
```diff
@@ -1,34 +1,46 @@
...
   end
 
   def call
+    if participant_declaration.paid?
+      make_awaiting_clawback
+    else
+      make_voided
+    end
+
+    Api::V1::ParticipantDeclarationSerializer.new(participant_declaration).serializable_hash.to_json
```
Comment (asmega) [inline app/services/void_participant_declaration.rb:17]: while we're here any chance we can make this return `participant_declaration` instead and let the controller decide what serialiser to use?

File: app/services/void_participant_declaration.rb [hunk 2]
```diff
@@ -1,34 +1,46 @@
...
 
 private
 
   def line_item
-    declaration.statement_line_items.find_by(state: %w[eligible payable])
-  end
-
-  def declaration
-    @declaration ||= ParticipantDeclaration.for_lead_provider(cpd_lead_provider).find(id)
+    participant_declaration.statement_line_items.find_by(state: %w[eligible payable])
```
Comment (asmega) [inline app/services/void_participant_declaration.rb:44]: on a side note possible bug there. this should probably cover `submitted` too

**Review by asmega — APPROVED** [review]

**Comment by StupidCodeFactory** [comment]
@mooktakim  I think this acceptance criteria isn't fulfilled, would you mind adding a test that covers this acceptance? 

`A provider can post a declaration of a type already associated with the participant and course and the declaration be eligible, providing the existing declaration is awaiting_clawback or clawed_back state`

**Comment by StupidCodeFactory** [comment]
I'd also expect to see a few test around the change schedule post clawback, if you could also add those that'd be wonderful, thanks.

File: spec/requests/api/v1/participant_declarations_spec.rb [hunk 3]
```diff
@@ -375,6 +375,48 @@ def build_params(attributes)
...
+                cpd_lead_provider:,
+                participant_profile: ect_profile,
+                course_identifier: valid_params[:course_identifier],
+              )
+            end
+
+            it "returns 200" do
+              params = build_params(valid_params)
+              post "/api/v1/participant-declarations", params: params
+              expect(response.status).to eq 200
```
Comment (StupidCodeFactory) [inline spec/requests/api/v1/participant_declarations_spec.rb:396]: @mooktakim here you want to assert that the a new declaration has been created and that it also is in the correct state alongside with a statement line item in the correct test.

File: spec/requests/api/v1/participant_declarations_spec.rb [hunk 4]
```diff
@@ -375,6 +375,87 @@ def build_params(attributes)
...
+            expect(ParticipantDeclaration.first).to be_awaiting_clawback
+
+            params = build_params(valid_params)
+            post "/api/v1/participant-declarations", params: params
+            expect(response.status).to eq 200
+
+            declarations = ParticipantDeclaration.all
+            expect(declarations.count).to eq(2)
+            expect(declarations[0]).to be_awaiting_clawback
+            expect(declarations[1]).to be_submitted
```
Comment (StupidCodeFactory) [inline spec/requests/api/v1/participant_declarations_spec.rb:403]: I'd personally prefer something a tad more explicit:

```ruby
params = build_params(valid_params)
expect {
  post "/api/v1/participant-declarations", params: params
  expect(response.status).to eq 200
}.to change { ect_profile.participant_declaration.submitted.count }.from(0).to(1)

# sanity check that the initial declaration is still on the correct state
expect(existing_participant_declaration).to be_awaiting_clawback
```

File: spec/requests/api/v1/participant_declarations_spec.rb [hunk 5]
```diff
@@ -375,6 +375,87 @@ def build_params(attributes)
...
+
+          it "returns 200" do
+            expect(ParticipantDeclaration.count).to eq(1)
+            expect(ParticipantDeclaration.first).to be_awaiting_clawback
+
+            params = build_params(valid_params)
+            post "/api/v1/participant-declarations", params: params
+            expect(response.status).to eq 200
+
+            declarations = ParticipantDeclaration.all
```
Comment (StupidCodeFactory) [inline spec/requests/api/v1/participant_declarations_spec.rb:400]: the ordering isn't guaranteed that's usually yields flaky specs

File: spec/requests/api/v1/participant_declarations_spec.rb [hunk 6]
```diff
@@ -375,6 +375,87 @@ def build_params(attributes)
...
+              :awaiting_clawback,
+              user: ect_profile.user,
+              cpd_lead_provider:,
+              participant_profile: ect_profile,
+              course_identifier: valid_params[:course_identifier],
+            )
+          end
+
+          it "returns 200" do
+            expect(ParticipantDeclaration.count).to eq(1)
```
Comment (StupidCodeFactory) [inline spec/requests/api/v1/participant_declarations_spec.rb:393]: To add an extra layer of assurance you always want to be a bit more precise in those types of assertions and you achieve those two lines in one assertion as well:

```ruby
expect(ect_profile.participant_profiles.awaiting_clawback.count).to eq(1)

File: spec/requests/api/v1/participant_declarations_spec.rb [hunk 7]
```diff
@@ -375,6 +375,87 @@ def build_params(attributes)
...
+            expect(ParticipantDeclaration.first).to be_clawed_back
+
+            params = build_params(valid_params)
+            post "/api/v1/participant-declarations", params: params
+            expect(response.status).to eq 200
+
+            declarations = ParticipantDeclaration.all
+            expect(declarations.count).to eq(2)
+            expect(declarations[0]).to be_clawed_back
+            expect(declarations[1]).to be_submitted
```
Comment (StupidCodeFactory) [inline spec/requests/api/v1/participant_declarations_spec.rb:430]: same remark as above ^^

File: spec/requests/api/v1/participant_declarations_spec.rb (hunk 6 above)
Comment (StupidCodeFactory) [reply spec/requests/api/v1/participant_declarations_spec.rb:393]: Those could be passing but not not for the reason you think, another factory may have created that record as a side effect, which happens a lot with FactoryBot when you're not careful.

File: app/services/void_participant_declaration.rb [hunk 8]
```diff
@@ -1,34 +1,46 @@
 # frozen_string_literal: true
 
 class VoidParticipantDeclaration
-  attr_accessor :cpd_lead_provider, :id
+  attr_reader :participant_declaration
```
Comment (StupidCodeFactory) [inline app/services/void_participant_declaration.rb:4]: This should be private, we're not accessing it from anywhere

File: spec/services/void_participant_declaration_spec.rb [hunk 9]
```diff
@@ -9,14 +9,13 @@
...
   let(:cpd_lead_provider) { profile.school_cohort.school.partnerships[0].lead_provider.cpd_lead_provider }
-  let(:another_cpd_lead_provider) { create(:cpd_lead_provider) }
 
   before do
     create(:partnership, school:)
   end
 
   describe "#call" do
-    let(:declaration) do
+    let(:participant_declaration) do
```
Comment (StupidCodeFactory) [inline spec/services/void_participant_declaration_spec.rb:18]: 🎖️ thanks for renaming this properly

**Review by StupidCodeFactory — APPROVED** [review]

**Pull request merged**